launcher/firstLaunch/firstlaunch_moc.cpp

```diff
@@ -338,248 +402,317 @@ void FirstLaunchView::extractGogData()
-		return;
-	}
+        overlay->deleteLater();
+        QMessageBox::critical(self, QObject::tr("Heroes III data not found!"), QObject::tr("No matching files found in the selected folder."));
+        return false;
+    }
 
-	QDir sourceData = sourceRoot.filePath(dirData.front());
-	QStringList roeFiles = sourceData.entryList({"*.lod"}, QDir::Filter::Files);
-	QStringList sodFiles = sourceData.entryList({"H3ab*.lod"}, QDir::Filter::Files);
-	QStringList hdFiles = sourceData.entryList({"*.pak"}, QDir::Filter::Files);
+    // 4) Copy with progress
+    overlay->setTitle(QObject::tr("Importing Heroes III data..."));
+    overlay->setIndeterminate(false);
+    overlay->setRange(plan.size());
 
-	if(sodFiles.empty())
+    for(int i = 0; i < plan.size(); ++i)
 	{
-		if (roeFiles.empty())
-		{
-			// Directory structure is correct (Data/Maps/Mp3) but no .lod archives that should be present in any install
-			QMessageBox::critical(this, tr("Heroes III data not found!"), noDataMessage);
-			return;
-		}
+        overlay->setFileName(QFileInfo(plan[i].dst).fileName());
+        overlay->setValue(i + 1);
+        qApp->processEvents();
 
-		if (!hdFiles.empty())
-		{
-			// HD Edition contains only RoE data so we can't use even unmodified files from it
-			QMessageBox::critical(this, tr("Heroes III data not found!"), tr("Heroes III: HD Edition files are not supported by VCMI.\nPlease select the directory with Heroes III: Complete Edition or Heroes III: Shadow of Death."));
-			return;
-		}
+        if (QFile::exists(plan[i].dst))
+            QFile::remove(plan[i].dst);
 
-		// RoE or some other unsupported edition. Demo version?
-		QMessageBox::critical(this, tr("Heroes III data not found!"), tr("Unknown or unsupported Heroes III version found.\nPlease select the directory with Heroes III: Complete Edition or Heroes III: Shadow of Death."));
-		return;
-	}
+        Helper::performNativeCopy(plan[i].src, plan[i].dst);
 
-	QStringList copyDirectories;
+        logGlobal->info("Copying '%s' -> '%s'", plan[i].src.toStdString(), plan[i].dst.toStdString());
+    }
 
-	copyDirectories += dirData.front();
-	if (!dirMaps.empty())
-		copyDirectories += dirMaps.front();
+    // 5) Optional cleanup
+    if(removeSource)
+        QDir(path).removeRecursively();
 
-	if (!dirMp3.empty())
-		copyDirectories += dirMp3.front();
+    overlay->deleteLater();
+    return true;
+}
 
-	QDir targetRoot = pathToQString(VCMIDirs::get().userDataPath());
+void FirstLaunchView::extractGogDataAsync(QString filePathBin, QString filePathExe)
+{
+    logGlobal->info("Extracting gog data from '%s' and '%s'", filePathBin.toStdString(), filePathExe.toStdString());
 
-	for(const QString & dirName : copyDirectories)
-	{
-		QDir sourceDir = sourceRoot.filePath(dirName);
-		QDir targetDir = targetRoot.filePath(dirName);
+#ifdef ENABLE_INNOEXTRACT
+    //  Show overlay immediately so the UI doesn't look frozen
+    auto *overlay = createOverlay(this, tr("Checking installer..."), true);
```

Comment:
what's the point of creating the overlay outside the actual handler? If you move it there, then you can use `QScopedPointer` to avoid manual call to `deleteLater`, as the handler is synchronous.